url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/MOVEM.java

```diff
@@ -0,0 +1,692 @@
+//       following disclaimer.
+//    o  Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//       following disclaimer in the documentation and/or other materials provided with the distribution.
+//    o  Neither the name of the M68k Project nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+*/
+
+// when the ea is specified by a pre-decreasing mode (eg movem.l range,-(an) , then, if the addressing register
+// itself is part of the range and for a 68000 and 68010, the value stored is the INITIAL value of the addressing reg.
+// For a 68020 and up, the value written is the initial value minus the size (word or long) of the operation.
+
+
+public class MOVEM implements InstructionHandler
+{
+	protected final Cpu cpu;
+
+	private static final int[] M2R_Timing = {
+			0, 0, 12, 12, 0, 16, 18, 16, 20, 16, 18
+	};
+	private static final int[] R2M_Timing = {
+			0, 0, 8, 0, 8, 12, 14, 12, 16
+	};
+
+	public MOVEM(Cpu cpu)
+	{
+		this.cpu = cpu;
+	}
+
+	public final void register(InstructionSet is)
+	{
+		int base;
+		Instruction i;
+
+		// reg to mem
+		for(int sz = 0; sz < 2; sz++)
+		{
+			if(sz == 0)
+			{
+				base = 0x4880;
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return movem_word_r2m(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Word, true);
+					}
+				};
+			}
+			else
+			{
+				base = 0x48c0;
+				i = new Instruction() {
+					public int execute(int opcode)
+					{
+						return movem_long_r2m(opcode);
+					}
+					public DisassembledInstruction disassemble(int address, int opcode)
+					{
+						return disassembleOp(address, opcode, Size.Long, true);
+					}
+				};
+			}
+
+			for(int ea_mode = 2; ea_mode < 8; ea_mode++)
+			{
+				if(ea_mode == 3)
+					continue;
+
+				for(int ea_reg = 0; ea_reg < 8; ea_reg++)
+				{
+					if(ea_mode == 7 && ea_reg > 1)
+						break;
+
+					is.addInstruction(base + (ea_mode << 3) + ea_reg, i);
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 88:100, 135:147.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.